src/Raven.Client/Documents/BulkInsert/BulkInsertOperation.cs

```diff
@@ -371,42 +458,51 @@ public async Task StoreAsync(object entity, string id, IMetadataDictionary metad
 
                 EndPreviousCommandIfNeeded();
 
-                try
-                {
-                    if (_first == false)
-                    {
-                        WriteComma();
-                    }
-
-                    _first = false;
-                    _inProgressCommand = CommandType.None;
-
-                    _currentWriter.Write("{\"Id\":\"");
-                    WriteString(id);
-                    _currentWriter.Write("\",\"Type\":\"PUT\",\"Document\":");
+                await WriteToStreamAsync(entity, id, metadata, CommandType.PUT).ConfigureAwait(false);
+            }
+        }
 
-                    _currentWriter.Flush();
+        private async ValueTask WriteToStreamAsync(object entity, string id, IMetadataDictionary metadata, CommandType type)
+        {
+            try
+            {
+                if (_first == false)
+                {
+                    WriteComma();
+                }
 
-                    if (_customEntitySerializer == null || _customEntitySerializer(entity, metadata, _currentWriter) == false)
-                    {
-                        using (var json = _conventions.Serialization.DefaultConverter.ToBlittable(entity, metadata, _context, _defaultSerializer))
-                            await json.WriteJsonToAsync(_currentWriter.BaseStream, _token).ConfigureAwait(false);
-                    }
+                _first = false;
+                _inProgressCommand = CommandType.None;
 
-                    _currentWriter.Write('}');
+                _currentWriter.Write("{\"Id\":\"");
+                WriteString(id);
+                _currentWriter.Write("\",\"Type\":\"");
+                _currentWriter.Write(type);
+                _currentWriter.Write("\",\"Document\":");
 
-                    await FlushIfNeeded().ConfigureAwait(false);
+                _currentWriter.Flush();
 
-                }
-                catch (Exception e)
+                if (_customEntitySerializer == null || _customEntitySerializer(entity, metadata, _currentWriter) == false)
                 {
-                    await HandleErrors(id, e).ConfigureAwait(false);
+                    using (var json = _conventions.Serialization.DefaultConverter.ToBlittable(entity, metadata, _context, _defaultSerializer))
+                        await json.WriteJsonToAsync(_currentWriter.BaseStream, _token).ConfigureAwait(false);
                 }
+
+                _currentWriter.Write('}');
+
+                await FlushIfNeeded().ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                await HandleErrors(id, e).ConfigureAwait(false);
             }
         }
-        
+
         private async Task HandleErrors(string documentId, Exception e)
         {
+            if (e.TargetSite.Name.Equals("ThrowAlreadyRunningTimeSeries"))
```

Comment:
let's fix it differently.

Introduce abstract BulkInsertClientException
Introduce BulkInsertInvalidOperationExcepton

throw in TS the BulkInsertInvalidOperationExcepton

check here is exception is BulkInsertClientException and rethrow

Reply:
Do we want all bulkInsert exceptions to be BulkInsertClientException, like BulkInsertProtocolViolationException, and BulkInsertAbortedException?

Reply:
When those exceptions are thrown, we want to download the exception from the server or just immediately throw? The answer to that Q is he answer to your question.

Reply:
done